File: src/utils/landingContent.ts
import { LandingContent } from '../components/landing/LandingMobileView';

type LandingEventLike = {
  title?: string | null;
  event_type?: string | null;
  date?: string | null;
  city?: string | null;
  venue_name?: string | null;
  venue_city?: string | null;
};

type LandingPageLike = {
  published?: boolean | null;
  content_json?: string | null;
  title?: string | null;
  date?: string | null;
  location?: string | null;
  cover_image?: string | null;
  message?: string | null;
  dress_code?: string | null;
};

const DEFAULT_SCHEDULE = [
  { time: '18:00', activity: 'Cocteil de bun venit' },
  { time: '19:00', activity: 'Ceremonia' },
  { time: '20:00', activity: 'Cina festiva' },
  { time: '22:00', activity: 'Petrecere' },
];

export const buildDefaultLandingContent = (event: LandingEventLike | null): LandingContent => ({
  published: false,
  coverImage: '',
  galleryImages: [],
  title: event?.event_type || event?.title || 'Evenimentul nostru',
  subtitle: 'Va invitam sa sarbatorim impreuna',
  date: event?.date || '',
  time: '18:00',
  location: event?.venue_name || event?.venue_city || event?.city || '',
  locationAddress: '',
  heroMessage: 'Cu drag va asteptam sa fiti alaturi de noi in aceasta zi speciala!',
  dateDescription: 'Pastrati data in calendar si pregatiti-va pentru o seara memorabila alaturi de noi.',
  detailsDescription: 'Vom reveni cu toate detaliile importante pentru ca voi sa va bucurati de eveniment fara griji.',
  dressCode: 'Tinuta eleganta',
  schedule: DEFAULT_SCHEDULE,
  timingIcons: [],
  colorPrimary: '#960010',
  colorSecondary: '#ec4899',
});

export const normalizeLandingContent = (
  value: Partial<LandingContent> | null | undefined,
  event: LandingEventLike | null
): LandingContent => {
  const defaults = buildDefaultLandingContent(event);
  const legacyMessage = (value as Partial<LandingContent> & { message?: string } | null | undefined)?.message;

  return {
    ...defaults,
    ...value,
    heroMessage: value?.heroMessage ?? legacyMessage ?? defaults.heroMessage,
    dateDescription: value?.dateDescription ?? legacyMessage ?? defaults.dateDescription,
    detailsDescription: value?.detailsDescription ?? legacyMessage ?? defaults.detailsDescription,
    published: Boolean(value?.published),
    galleryImages: Array.isArray(value?.galleryImages) ? value.galleryImages : defaults.galleryImages,
    schedule: Array.isArray(value?.schedule) ? value.schedule : defaults.schedule,
    timingIcons: Array.isArray(value?.timingIcons) ? value.timingIcons : defaults.timingIcons,
  };
};

export const parseLandingContentJson = (
  contentJson: string | null | undefined,
  event: LandingEventLike | null
): LandingContent | null => {
  if (!contentJson) {
    return null;
  }

  try {
    return normalizeLandingContent(JSON.parse(contentJson), event);
  } catch (error) {
    console.error('Failed to parse landing content from backend:', error);
    return null;
  }
};

export const buildLandingContentFromApi = (
  landingPage: LandingPageLike,
  event: LandingEventLike | null
): LandingContent => {
  return (
    parseLandingContentJson(landingPage.content_json, event) ||
    normalizeLandingContent(
      {
        published: Boolean(landingPage.published),
        title: landingPage.title || undefined,
        date: landingPage.date || undefined,
        location: landingPage.location || undefined,
        coverImage: landingPage.cover_image || undefined,
        heroMessage: landingPage.message || undefined,
        dateDescription: landingPage.message || undefined,
        detailsDescription: landingPage.message || undefined,
        dressCode: landingPage.dress_code || undefined,
      },
      event
    )
  );
};

File: backend/app/services/landing_content.py
import json
from typing import Any, Dict, Optional

from app.models.landing import LandingPage


def parse_landing_content(content_json: Optional[str]) -> Dict[str, Any]:
    if not content_json:
        return {}

    try:
        parsed = json.loads(content_json)
    except (TypeError, ValueError):
        return {}

    return parsed if isinstance(parsed, dict) else {}


def build_legacy_content_snapshot(landing_page: LandingPage) -> dict[str, Any]:
    return {
        "published": bool(landing_page.published),
        "coverImage": landing_page.cover_image or "",
        "galleryImages": [],
        "title": landing_page.title or "",
        "subtitle": "",
        "date": landing_page.date or "",
        "time": "",
        "location": landing_page.location or "",
        "locationAddress": "",
        "heroMessage": landing_page.message or "",
        "dateDescription": landing_page.message or "",
        "detailsDescription": landing_page.message or "",
        "dressCode": landing_page.dress_code or "",
        "schedule": [],
        "timingIcons": [],
        "colorPrimary": "#960010",
        "colorSecondary": "#ec4899",
    }


def sync_landing_fields_from_content(landing_page: LandingPage) -> None:
    content = parse_landing_content(landing_page.content_json)
    if not content:
        content = build_legacy_content_snapshot(landing_page)

    landing_page.content = content
    landing_page.title = content.get("title") or landing_page.title
    landing_page.date = content.get("date") or landing_page.date
    landing_page.location = content.get("location") or landing_page.location
    landing_page.cover_image = content.get("coverImage") or landing_page.cover_image
    landing_page.message = content.get("heroMessage") or content.get("message") or landing_page.message
    landing_page.dress_code = content.get("dressCode") or landing_page.dress_code
    landing_page.published = bool(content.get("published", landing_page.published))
    landing_page.content_json = json.dumps(content)


def update_landing_from_payload(landing_page: LandingPage, payload: dict[str, Any]) -> None:
    content_json = payload.pop("content_json", None)

    if content_json is not None:
        landing_page.content_json = content_json

    for key, value in payload.items():
        setattr(landing_page, key, value)

    sync_landing_fields_from_content(landing_page)
